### PR TITLE
PHP8 compatibility update

### PR DESCRIPTION
**Pull request by kohlerdominik**
**DO NOT MERGE! READ FIRST**

This PR adds compatibility to PHP8 to run tests in Travis.

As Issue #72 states, currently not all dependencies are updated. In particular we are waiting for [this PR](https://github.com/khanamiryan/php-qrcode-detector-decoder/pull/103).

So this [commit](https://github.com/sprain/php-swiss-qr-bill/pull/106/commits/43880bb93457f2287095533c18026e5ffddc882d) from the PR overwrites the blocking dependencies with fixed versions. It must be reverted before the PR gets merged!

- [endroid/qr-code](https://github.com/endroid/qr-code): The master branch is updated to support PHP8. However, due missing compatibility from khanamiryan/qrcode-detector-decoder, there's no release yet.
- [khanamiryan/qrcode-detector-decoder](https://github.com/khanamiryan/php-qrcode-detector-decoder): has the open PR with the fixes to PHP8 mentioned above.

### CODE REVIEW

**Comment by kohlerdominik** [comment]
@sprain Tests are green, except with PHP7.2, as it's no longer supported in endroid/qr-code. 
PHP 7.2 is no longer supported, so can we drop that in the next release?

**Comment by sprain** [comment]
> PHP 7.2 reached end of support, so can we drop that in the next release?

Yes, we can.
Thanks a lot for your work here! I'll be happy to have a look asap.

**Comment by sprain** [comment]
Sorry for the delay here … very busy with other stuff right now. Will try to get back asap.

**Comment by kohlerdominik** [comment]
Hi Manuel

no worries, still waiting for the khanamiryan/qrcode-detector-decoder anyways...

File: composer.json [hunk 1]
```diff
@@ -3,13 +3,19 @@
...
     "require": {
-        "php": "^7.2",
+        "php": "^7.3|^8.0",
         "symfony/validator": "^3.4.31|^4.2|^5.0",
         "symfony/intl": "^3.4|^4.2|^5.0",
-        "khanamiryan/qrcode-detector-decoder": "^1.0.3",
+        "khanamiryan/qrcode-detector-decoder": "dev-issue-php8 as 1.0.5",
         "kmukku/php-iso11649": "^1.5",
-        "endroid/qr-code": "^3.9.1"
+        "endroid/qr-code": "dev-master"
```
Comment (PATROMO) [inline composer.json:18]: v4.0.0 is released. https://github.com/endroid/qr-code/releases/tag/4.0.0

**Comment by kohlerdominik** [comment]
Yes, I think it's your best shot to migrate to the new version. I did a quick test, and I think there are a couple of changes neccessary:

- We have a custom implementation of `Sprain\SwissQrBill\QrCode\QrCode`, which extends `Endroid\QrCode\QrCode`. But the extended class is now `final`, so it's necessary to devcide what to do with this overrides.
- Endroid removed the former `QrCode`-class and instead provides a `Builder` with a different interfaces. It requires adjustments along the whole QrCode-Generation process. Outputs seems the same though, so i think the `AbstractOutput` and it's Subclasses need only minor adjustments.
- From what I can see, we do not actually utilize any features from khanamiryan/qrcode-detector-decoder, so with the package being optional now, we can completely remove it from our dependencies. I'm not sure why it's on there, though.

@sprain can you provide a quick feedback to this conclusions?
@PATROMO As it looks like that's neither for Manuel nor for me the PHP8 update is very urgent, it might take some time until this is resolved. But if you'd like to make a PR, I think it's very welcome.

**Comment by sprain** [comment]
Thanks a lot for pushing this forward, @kohlerdominik! Currently clients are keeping me busy with other stuff :) 

> We have a custom implementation of Sprain\SwissQrBill\QrCode\QrCode, which extends Endroid\QrCode\QrCode. But the extended class is now final, so it's necessary to devcide what to do with this overrides.

At first glimpse it actually looks like Endroid has a nicer, decoupled setup now. Extending the class might become unnecessary.

> It requires adjustments along the whole QrCode-Generation process. Outputs seems the same though, so i think the AbstractOutput and it's Subclasses need only minor adjustments.

I suppose the `Outputs` will now need to be aware of the Endroid `Writer`. Looks like some refactoring will be required.

> From what I can see, we do not actually utilize any features from khanamiryan/qrcode-detector-decoder

It's used in `QrBillTest.php` (and therefore should be a dev-dependency only). Maybe there is an alternative out there …

> As it looks like that's neither for Manuel nor for me the PHP8 update is very urgent, it might take some time until this is resolved. But if you'd like to make a PR, I think it's very welcome.

Indeed it's not urgent for me, but of course a goal. I hope to be able to dig into into soon. Any further help is very welcome!

**Comment by sprain** [comment]
**UPDATE**
As this pr is not yet super important for us at work, I am limited to looking at it during my free time. As you can see, those resources are limited.

However, I have been granted early access to the Github's new one-time sponsoring program. That provides a good opportunity to try something new!

**So here is the deal:** [You can now support work on this library with a one-time sponsoring.](https://github.com/sponsors/sprain?frequency=one-time&sponsor=sprain) I promise that any one-time sponsoring coming through this channel will be applied to working on this library, with priority on this pull request.

This is valid up to a total amount of $7200 (or 40 hours of work). The work will be put in as soon as possible and at my best possible quality level. However, it is still all based on voluntariness, as a sponsorship through Github does not provide the necessary legal framework to create a contract. Therefore no mutual obligations arise from a sponsorship. You'll have to trust me 😉

**Comment by sprain** [comment]
Before moving on to PHP8, let's add some other changes. You are welcome to have a look at #121, @kohlerdominik.

**Comment by kohlerdominik** [comment]
> 
> 
> Before moving on to PHP8, let's add some other improvements. You are welcome to have a look at #121, @kohlerdominik.

I'll try to find some time until sunday evening.

**Comment by sprain** [comment]
> I'll try to find some time until sunday evening.

No worries, only if you want to. I‘ll be on vacation for the next two weeks and will leave it unmerged until I get back.

**Comment by kohlerdominik** [comment]
Good news everybody! khanamiryan/php-qrcode-detector-decoder#103 just got merged. So we can probably release a php8 compatible version before the switching to endroid/qr-code v4

**Pull request closed without merging**